tools/build_langserver/langserver/analyzer.go

```diff
@@ -205,6 +214,101 @@ func (a *Analyzer) AspStatementFromFile(uri lsp.DocumentURI) ([]*asp.Statement,
 	return stmts, nil
 }
 
+// AspStatementFromContent returns a slice of asp.Statement given content string(usually workSpaceStore.doc.TextInEdit)
+func (a *Analyzer) AspStatementFromContent(content string) []*asp.Statement {
+	byteContent := []byte(content)
+
+	stmts, err := a.parser.ParseData(byteContent, "")
+	if err != nil {
+		log.Warning("reading only partial of the file due to parsing failure: %s ", err)
+	}
+
+	return stmts
+}
+
+// IdentsFromContent returns a channel of Identifier object
+func (a *Analyzer) IdentsFromContent(content string) chan *Identifier {
+	stmts := a.AspStatementFromContent(content)
+
+	ch := make(chan *Identifier)
+	go func() {
+		for _, stmt := range stmts {
+			if stmt.Ident != nil {
+				ident := a.identFromStatement(stmt)
+				ch <- ident
+			}
+		}
+		close(ch)
+	}()
+
+	return ch
+}
+
+// FuncCallFromContentAndPos returns a Identifier object represents function call,
+// Only returns the not nil object when the Identifier is within the range specified by the position
+func (a *Analyzer) FuncCallFromContentAndPos(content string, pos lsp.Position) *Identifier {
+	idents := a.IdentsFromContent(content)
+
+	for i := range idents {
+		if i.Type != "call" {
+			continue
+		}
+
+		if withInRangeLSP(i.Pos, i.EndPos, pos) {
+			return i
+		}
+	}
+
+	return nil
+}
+
+// VariablesFromContent returns a map of variable name to Variable
+func (a *Analyzer) VariablesFromContent(content string) map[string]Variable {
+	idents := a.IdentsFromContent(content)
+
+	vars := make(map[string]Variable)
+	for i := range idents {
+		if i.Type != "assign" && i.Type != "augAssign" {
+			continue
+		}
+
+		var varType string
+		if i.Type == "assign" {
+			varType = getVarType(i.Action.Assign.Val)
+		} else if i.Type == "augAssign" {
+			varType = getVarType(i.Action.AugAssign.Val)
+		}
+
+		if varType != "" {
+			variable := Variable{
+				Name: i.Name,
+				Type: varType,
+			}
+			vars[i.Name] = variable
+		}
+	}
+
+	return vars
+}
+
+func getVarType(valExpr *asp.ValueExpression) string {
+	v := reflect.ValueOf(valExpr).Elem()
```

Comment:
I'm a bit dubious that reflection is the best thing to use here? Do you really want to access any possible string field out of the ValueExpression (as opposed to for example reading the string / fstring fields?)

Reply:
So basically, I would need only the type of the variable here, so I can later use it to figure out the correct completion to return. for example, if I have a variable: `blah = "my string"`. later in the file, when I use `blah.`, then the LPS will need to know what type of variable `blah` is. I mean I could also do `if, else if` statement to see which field of `asp.ValueExpression` is not `nil`. I feel like it's a bit tedious. Unless we have any other way of figuring out the value of the variable(`*asp.Identifier`)? Any thoughts?

Reply:
Yeah but you aren't reading the type of variable, you're reading what the name is in the struct, which includes things that aren't types (like `Ident` or `Lambda` and things like that).